### PR TITLE
ref(controller): remove initial from .deploy()

### PR DESCRIPTION
**Pull request by bacongobbler**
initial is only equal to the time when the structure has no values, so
it makes no sense to check for initial in any other case.

### CODE REVIEW

**Comment by carmstrong** [comment]
I like the cleanup, but worried there may be a hidden regression here. Do we have adequate test coverage for these changes?

**Comment by bacongobbler** [comment]
I found that there was a regression with releases:rollback, but the tests caught that and line 333 was born. I guess I could double-check all calls to `app.deploy()` and see if we should be catching that error in other places, but otherwise we should be good.

**Comment by bacongobbler** [comment]
On second thought, it might be a logical error in my code. Let me check that.

**Comment by bacongobbler** [comment]
much better.

**Comment by carmstrong** [comment]
> much better.

What was the bug? Did tests cover it?

**Comment by bacongobbler** [comment]
The tests covered it, but basically what was occurring was the following:

```
><> venv/bin/python manage.py test api.ConfigTest.test_config_set_unicode
WARNING Cannot synchronize with etcd cluster
Creating test database for alias 'default'...
E
======================================================================
ERROR: test_config_set_unicode (api.tests.test_config.ConfigTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/venv/local/lib/python2.7/site-packages/mock.py", line 1201, in patched
    return func(*args, **keywargs)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/api/tests/test_config.py", line 172, in test_config_set_unicode
    HTTP_AUTHORIZATION='token {}'.format(self.token))
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/venv/local/lib/python2.7/site-packages/django/test/client.py", line 483, in post
    response = super(Client, self).post(path, data=data, content_type=content_type, **extra)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/venv/local/lib/python2.7/site-packages/django/test/client.py", line 302, in post
    return self.request(**r)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/venv/local/lib/python2.7/site-packages/django/test/client.py", line 444, in request
    six.reraise(*exc_info)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/venv/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 112, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/venv/local/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 57, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/venv/local/lib/python2.7/site-packages/rest_framework/viewsets.py", line 85, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/venv/local/lib/python2.7/site-packages/rest_framework/views.py", line 407, in dispatch
    response = self.handle_exception(exc)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/venv/local/lib/python2.7/site-packages/rest_framework/views.py", line 404, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/api/views.py", line 84, in create
    return super(AppResourceViewSet, self).create(request, **kwargs)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/api/views.py", line 61, in create
    return super(BaseDeisViewSet, self).create(request, *args, **kwargs)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/venv/local/lib/python2.7/site-packages/rest_framework/mixins.py", line 21, in create
    self.perform_create(serializer)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/api/viewsets.py", line 21, in perform_create
    self.post_save(obj)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/api/views.py", line 193, in post_save
    config.app.deploy(self.request.user, self.release)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/api/models.py", line 347, in deploy
    self._default_scale(user, release)
  File "/home/bacongobbler/go/src/github.com/deis/deis/controller/api/models.py", line 352, in _default_scale
    if not release.build.sha:
AttributeError: 'NoneType' object has no attribute 'sha'

----------------------------------------------------------------------
Ran 1 test in 0.079s

FAILED (errors=1)
Destroying test database for alias 'default'...
```

By default, initial was set to `False`. Since we deploy [when configuration changes](https://github.com/deis/deis/blob/master/controller/api/views.py#L193), we'd go through the steps to deploy. However, at this point in the tests there's no Build present for the release (i.e. `deis create` into `deis config:set`). `_default_scale()` would be called because no structure existed for this app yet (no scaling operation was done at this time) which would then fail because we're calling `.sha` on a NoneType build object. Checking for both a non-null structure and a non-null build is equivalent to the behaviour it was expecting.

TL;DR our API test coverage is pretty extensive (and awesome).

File: controller/api/models.py [hunk 1]
```diff
@@ -343,7 +343,7 @@ def deploy(self, user, release, initial=False):
             self._destroy_containers(existing)
 
         # perform default scaling if necessary
-        if initial:
+        if self.structure == {} and release.build is not None:
```
Comment (carmstrong) [inline controller/api/models.py]: Looks like `and release.build is not None` is additional logic that we didn't have before - is this intentional?

Comment (bacongobbler) [inline controller/api/models.py]: This is intentional so that users who do a `deis config:set` before they do an intial deploy won't blow up on them. If you remove that call, you'll end up with the errors listed above. This is because `_default_scale` requires that the release has a non-null build object.

**Comment by carmstrong** [comment]
Code LGTM with @mboersma's eyes as well.

**Comment by mboersma** [comment]
Code LGTM.

**Comment by bacongobbler** [comment]
We talked about this over last week that the call where the config view calling app.deploy was weird and something we'd want to address by moving it back to the Release object. Do you guys want me to address that here in this PR, or is a followup ok?

**Comment by mboersma** [comment]
A separate PR is totes ok.
